File: Source/Game/Podium/StandardPodium.cs
using System;
using System.Collections.Generic;
using WCSharp.Api;
using WCSharp.Shared.Data;

/// <summary>
/// Players are going to be rewarded
/// </summary>
public static class StandardPodium
{
    private static Queue<(player Player, Point position)> PodiumQueue = new Queue<(player Player, Point position)>();
    private static List<unit> MovedUnits = new List<unit>();
    private static string PodiumType = "";
    private static string Color = Colors.COLOR_YELLOW_ORANGE;
    private const string HighestScore = "highest score";
    private const string MostSaves = "most saves";
    private const string HighestRatio = "highest ratio";
    private const string HighestStreak = "highest streak";

    public static void BeginPodiumActions()
    {
        PodiumUtil.SetCameraToPodium();
        Utility.SimpleTimer(3.0f, ProcessPodiumTypeActions);
    }

    private static void EnqueueTopScorePlayers()
    {
        var topScores = PodiumUtil.SortPlayersByScore();
        var podiumPositions = PodiumManager.PodiumSpots;
        for (int i = topScores.Count - 1; i >= 0; i--)
        {
            var player = topScores[i];
            var position = podiumPositions[i];
            PodiumQueue.Enqueue((player, position));
        }
        PodiumType = HighestScore;
    }

    private static void EnqueueTopSavesPlayers()
    {
        var topSaves = PodiumUtil.SortPlayersBySaves();
        var podiumPositions = PodiumManager.PodiumSpots;
        for (int i = topSaves.Count - 1; i >= 0; i--)
        {
            var player = topSaves[i];
            var position = podiumPositions[i];
            PodiumQueue.Enqueue((player, position));
        }
        PodiumType = MostSaves;
    }

    private static void EnqueueTopRatioPlayers()
    {
        var topRatios = PodiumUtil.SortPlayersByHighestRatio();
        var podiumPositions = PodiumManager.PodiumSpots;
        for (int i = topRatios.Count - 1; i >= 0; i--)
        {
            var player = topRatios[i];
            var position = podiumPositions[i];
            PodiumQueue.Enqueue((player, position));
        }
        PodiumType = HighestRatio;
    }

    private static void EnqueueTopStreakPlayers()
    {
        var topStreaks = PodiumUtil.SortPlayersByHighestSaveStreak();
        var podiumPositions = PodiumManager.PodiumSpots;
        for (int i = topStreaks.Count - 1; i >= 0; i--)
        {
            var player = topStreaks[i];
            var position = podiumPositions[i];
            PodiumQueue.Enqueue((player, position));
        }
        PodiumType = HighestStreak;
    }

    private static void ProcessNextPodiumAction()
    {
        if (PodiumQueue.Count == 0)
        {
            ProcessPodiumTypeActions();
            return;
        }
        var (player, position) = PodiumQueue.Dequeue();
        var kitty = Globals.ALL_KITTIES[player].Unit;
        kitty.SetPosition(position.X, position.Y);
        kitty.SetFacing(270);
        kitty.IsPaused = true;
        MovedUnits.Add(kitty);
        Console.WriteLine($"{Colors.PlayerNameColored(player)}{Color} earned {PodiumUtil.PlacementString(PodiumQueue.Count + 1)} place for {PodiumType} with {GetStatBasedOnType(player)}|r");
        Utility.SimpleTimer(5.0f, ProcessNextPodiumAction);
    }

    /// <summary>
    /// Topscore => Saves => Ratio => Streak .. End
    /// </summary>
    private static void ProcessPodiumTypeActions()
    {
        PodiumQueue.Clear();
        PodiumUtil.ClearPodiumUnits(MovedUnits);
        switch (PodiumType)
        {
            case "":
                EnqueueTopScorePlayers();
                break;

            case HighestScore:
                EnqueueTopSavesPlayers();
                break;

            case MostSaves:
                EnqueueTopRatioPlayers();
                break;

            case HighestRatio:
                EnqueueTopStreakPlayers();
                break;

            case HighestStreak:
                PodiumUtil.EndingGameThankyou();
                return;

            default:
                break;
        }
        ProcessNextPodiumAction();
    }

    private static string GetStatBasedOnType(player player)
    {
        var stats = Globals.ALL_KITTIES[player].CurrentStats;
        switch (PodiumType)
        {
            case HighestScore:
                return (stats.TotalSaves - stats.TotalDeaths).ToString();

            case MostSaves:
                return stats.TotalSaves.ToString();

            case HighestRatio:
                return stats.TotalDeaths == 0 ? stats.TotalSaves.ToString("F2") : ((float)stats.TotalSaves / stats.TotalDeaths).ToString("F2");

            case HighestStreak:
                return stats.MaxSaveStreak.ToString();

            default:
                return "n/a";
        }
    }
}

File: src/Game/Podium/StandardPodium.ts
import { Globals } from 'src/Global/Globals'
import { Colors } from 'src/Utility/Colors/Colors'
import { ColorUtils } from 'src/Utility/Colors/ColorUtils'
import { Queue } from 'src/Utility/Queue'
import { Utility } from 'src/Utility/Utility'
import { MapPlayer, Point, Unit } from 'w3ts'
import { PodiumUtil } from './PodiumUtil'

/// <summary>
/// Players are going to be rewarded
/// </summary>
export class StandardPodium {
    private static PodiumQueue = new Queue<[MapPlayer, Point]>()
    private static MovedUnits: Unit[] = []
    private static PodiumType: string = ''
    private static Color: string = Colors.COLOR_YELLOW_ORANGE
    private static HighestScore: string = 'score: highest'
    private static MostSaves: string = 'saves: most'
    private static HighestRatio: string = 'ratio: highest'
    private static HighestStreak: string = 'streak: highest'

    public static BeginPodiumActions = () => {
        PodiumUtil.SetCameraToPodium()
        Utility.SimpleTimer(3.0, StandardPodium.ProcessPodiumTypeActions)
    }

    private static EnqueueTopScorePlayers = () => {
        const topScores = PodiumUtil.SortPlayersByScore()
        const podiumPositions = PodiumUtil.PodiumSpots
        for (let i: number = topScores.length - 1; i >= 0; i--) {
            const player = topScores[i]
            const position = podiumPositions[i]
            StandardPodium.PodiumQueue.enqueue([player, position])
        }
        StandardPodium.PodiumType = StandardPodium.HighestScore
    }

    private static EnqueueTopSavesPlayers = () => {
        const topSaves = PodiumUtil.SortPlayersBySaves()
        const podiumPositions = PodiumUtil.PodiumSpots
        for (let i: number = topSaves.length - 1; i >= 0; i--) {
            const player = topSaves[i]
            const position = podiumPositions[i]
            StandardPodium.PodiumQueue.enqueue([player, position])
        }
        StandardPodium.PodiumType = StandardPodium.MostSaves
    }

    private static EnqueueTopRatioPlayers = () => {
        const topRatios = PodiumUtil.SortPlayersByHighestRatio()
        const podiumPositions = PodiumUtil.PodiumSpots
        for (let i: number = topRatios.length - 1; i >= 0; i--) {
            const player = topRatios[i]
            const position = podiumPositions[i]
            StandardPodium.PodiumQueue.enqueue([player, position])
        }
        StandardPodium.PodiumType = StandardPodium.HighestRatio
    }

    private static EnqueueTopStreakPlayers = () => {
        const topStreaks = PodiumUtil.SortPlayersByHighestSaveStreak()
        const podiumPositions = PodiumUtil.PodiumSpots
        for (let i: number = topStreaks.length - 1; i >= 0; i--) {
            const player = topStreaks[i]
            const position = podiumPositions[i]
            StandardPodium.PodiumQueue.enqueue([player, position])
        }
        StandardPodium.PodiumType = StandardPodium.HighestStreak
    }

    private static ProcessNextPodiumAction = () => {
        if (StandardPodium.PodiumQueue.length === 0) {
            StandardPodium.ProcessPodiumTypeActions()
            return
        }
        const [player, position] = StandardPodium.PodiumQueue.dequeue()!
        const kitty = Globals.ALL_KITTIES.get(player)!.Unit
        kitty.setPosition(position.x, position.y)
        kitty.setFacingEx(270)
        kitty.paused = true
        StandardPodium.MovedUnits.push(kitty)
        print(
            `${ColorUtils.PlayerNameColored(player)}${StandardPodium.Color} earned ${PodiumUtil.PlacementString(StandardPodium.PodiumQueue.length + 1)} for: place ${StandardPodium.PodiumType} with ${StandardPodium.GetStatBasedOnType(player)}|r`
        )
        Utility.SimpleTimer(5.0, StandardPodium.ProcessNextPodiumAction)
    }

    /// <summary>
    /// Topscore => Saves => Ratio => Streak .. End
    /// </summary>
    private static ProcessPodiumTypeActions = () => {
        StandardPodium.PodiumQueue.length = 0
        PodiumUtil.ClearPodiumUnits(StandardPodium.MovedUnits)
        switch (StandardPodium.PodiumType) {
            case '':
                StandardPodium.EnqueueTopScorePlayers()
                break

            case StandardPodium.HighestScore:
                StandardPodium.EnqueueTopSavesPlayers()
                break

            case StandardPodium.MostSaves:
                StandardPodium.EnqueueTopRatioPlayers()
                break

            case StandardPodium.HighestRatio:
                StandardPodium.EnqueueTopStreakPlayers()
                break

            case StandardPodium.HighestStreak:
                PodiumUtil.EndingGameThankyou()
                return

            default:
                break
        }
        StandardPodium.ProcessNextPodiumAction()
    }

    private static GetStatBasedOnType = (player: MapPlayer) => {
        const stats = Globals.ALL_KITTIES.get(player)!.CurrentStats
        switch (StandardPodium.PodiumType) {
            case StandardPodium.HighestScore:
                return (stats.TotalSaves - stats.TotalDeaths).toString()

            case StandardPodium.MostSaves:
                return stats.TotalSaves.toString()

            case StandardPodium.HighestRatio:
                return stats.TotalDeaths === 0
                    ? stats.TotalSaves.toFixed(2)
                    : (stats.TotalSaves / stats.TotalDeaths).toFixed(2)

            case StandardPodium.HighestStreak:
                return stats.MaxSaveStreak.toString()

            default:
                return 'n/a'
        }
    }
}
